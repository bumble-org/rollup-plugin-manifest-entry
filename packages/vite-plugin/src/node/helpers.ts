import { simple } from 'acorn-walk'
import { createHash as _hash } from 'crypto'
import debug from 'debug'
import fg from 'fast-glob'
import { AcornNode, PluginContext } from 'rollup'
import v8 from 'v8'
import type {
  ManifestV3,
  WebAccessibleResourceById,
  WebAccessibleResourceByMatch,
} from './manifest'
import type {
  AcornIdentifier,
  AcornLiteral,
  AcornMemberExpression,
  AcornTemplateElement,
  ManifestFiles,
} from './types'

export const _debug = (id: string) => debug('crx').extend(id)

export const structuredClone = <T>(obj: T): T => {
  return v8.deserialize(v8.serialize(obj))
}

export const createHash = (data: string, length = 5): string =>
  _hash('sha1')
    .update(data)
    .digest('base64')
    .replace(/[^A-Za-z0-9]/g, '')
    .slice(0, length)

export const isString = (x: unknown): x is string => typeof x === 'string'

type Falsy = false | 0 | '' | null | undefined
export const isTruthy = <T>(x: T | Falsy): x is T => !!x

export const isPresent = <T>(x: T | null | undefined): x is T =>
  x !== null && typeof x !== 'undefined'

export function isObject<T>(
  value: T,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
): value is Extract<T, Record<string, any>> {
  return Object.prototype.toString.call(value) === '[object Object]'
}

export const isResourceByMatch = (
  x: WebAccessibleResourceById | WebAccessibleResourceByMatch,
): x is WebAccessibleResourceByMatch => 'matches' in x

export async function manifestFiles(
  manifest: ManifestV3,
  options: fg.Options = {},
): Promise<ManifestFiles> {
  // JSON
  let locales: string[] = []
  if (manifest.default_locale)
    locales = await fg('_locales/**/messages.json', options)

  const rulesets =
    manifest.declarative_net_request?.rule_resources.flatMap(
      ({ path }) => path,
    ) ?? []

  const contentScripts = manifest.content_scripts?.flatMap(({ js }) => js) ?? []
  const contentStyles = manifest.content_scripts?.flatMap(({ css }) => css)
  const serviceWorker = manifest.background?.service_worker
  const htmlPages = htmlFiles(manifest)

  const icons = [
    Object.values(isString(manifest.icons) ? [manifest.icons] : manifest.icons ?? {}) as string[],
    Object.values(isString(manifest.action?.default_icon) ? [manifest.action?.default_icon] : manifest.action?.default_icon ?? {}) as string[],
  ].flat()

  let webAccessibleResources: string[] = []
  if (manifest.web_accessible_resources) {
    const resources = await Promise.all(
      manifest.web_accessible_resources
        .flatMap(({ resources }) => resources!)
        .map(async (r) => {
          // don't copy node_modules, etc
          if (['*', '**/*'].includes(r)) return undefined
          if (fg.isDynamicPattern(r)) return fg(r, options)
          return r
        }),
    )
    webAccessibleResources = resources.flat().filter(isString)
  }

  return {
    contentScripts: [...new Set(contentScripts)].filter(isString),
    contentStyles: [...new Set(contentStyles)].filter(isString),
    html: htmlPages,
    icons: [...new Set(icons)].filter(isString),
    locales: [...new Set(locales)].filter(isString),
    rulesets: [...new Set(rulesets)].filter(isString),
    background: [serviceWorker].filter(isString),
    webAccessibleResources,
  }
}

export function htmlFiles(manifest: ManifestV3): string[] {
  const files = [
    manifest.action?.default_popup,
    Object.values(manifest.chrome_url_overrides ?? {}),
    manifest.devtools_page,
    manifest.options_page,
    manifest.options_ui?.page,
    manifest.sandbox?.pages,
  ]
    .flat()
    .filter(isString)
    .map((s) => s.split('#')[0])
    .sort()
  return [...new Set(files)]
}

export function isMemberExpression(n: AcornNode): n is AcornMemberExpression {
  return n.type === 'MemberExpression'
}

export function isLiteral(n: AcornNode): n is AcornLiteral {
  return n.type === 'Literal'
}

export function isIdentifier(n: AcornNode): n is AcornIdentifier {
  return n.type === 'Identifier'
}

export function decodeManifest(this: PluginContext, code: string): ManifestV3 {
  const tree = this.parse(code)

  let literal: AcornLiteral | undefined
  let templateElement: AcornTemplateElement | undefined

  simple(tree, {
    Literal(node) {
      literal = node as AcornLiteral
    },
    TemplateElement(node) {
      templateElement = node as AcornTemplateElement
    },
  })

  let manifestJson: string | undefined = literal?.value
  if (!manifestJson) manifestJson = templateElement?.value?.cooked
  if (!manifestJson) throw new Error('unable to parse manifest code')

  let result: ManifestV3 = JSON.parse(manifestJson)
  if (typeof result === 'string') result = JSON.parse(result)
  return result
}

export function encodeManifest(manifest: ManifestV3): string {
  const json = JSON.stringify(JSON.stringify(manifest))
  return `export default ${json}`
}

/**
 * [Strip paths for `web_accessible_resources`'s `matches` Â· Issue
 * #282](https://github.com/crxjs/chrome-extension-tools/issues/282)
 */
export const stubMatchPattern = (pattern: string): string => {
  /**
   * Allow <all_urls> in matches section. 
   * [Issue #459](https://github.com/crxjs/chrome-extension-tools/issues/459)
   */
  if (pattern === "<all_urls>") {
    return pattern;
  }
  const [schema, rest] = pattern.split('://')
  const [origin, pathname] = rest.split('/')
  const root = `${schema}://${origin}`
  return pathname ? `${root}/*` : root
}
