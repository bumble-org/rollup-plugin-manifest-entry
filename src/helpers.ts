import { OutputAsset, OutputChunk, OutputBundle } from 'rollup'
import prettier from 'prettier'
import { CheerioFile } from './html-inputs'
import { ChromeExtensionManifest } from './manifest'

export const not = <T>(fn: (x: T) => boolean) => (x: T) => !fn(x)

export function isChunk(
  x: OutputChunk | OutputAsset,
): x is OutputChunk {
  return x.type === 'chunk'
}

export function isAsset(
  x: OutputChunk | OutputAsset,
): x is OutputChunk {
  return x.type === 'asset'
}

export function isString(x: any): x is string {
  return typeof x === 'string'
}

export function isJsonFilePath(x: any): x is string {
  return isString(x) && x.endsWith('json')
}

export const formatHtml = ($: CheerioFile) =>
  prettier.format($.html(), { parser: 'html' })

/**
 * Update the manifest source in the output bundle
 */
export const updateManifest = (
  updater: (
    manifest: ChromeExtensionManifest,
  ) => ChromeExtensionManifest,
  bundle: OutputBundle,
  handleError?: (message: string) => void,
): OutputBundle => {
  try {
    const manifestKey = 'manifest.json'
    const manifestAsset = bundle[manifestKey] as OutputAsset

    if (!manifestAsset) {
      throw new Error(
        'No manifest.json in the rollup output bundle.',
      )
    }

    const manifest = JSON.parse(
      manifestAsset.source as string,
    ) as ChromeExtensionManifest

    const result = updater(manifest)

    manifestAsset.source = JSON.stringify(result, undefined, 2)
  } catch (error) {
    if (handleError) {
      handleError(error.message)
    } else {
      throw error
    }
  }

  return bundle
}
