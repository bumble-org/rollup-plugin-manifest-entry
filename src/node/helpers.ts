import { simple } from 'acorn-walk'
import { createHash as _hash } from 'crypto'
import debug from 'debug'
import fg from 'fast-glob'
import { PluginContext } from 'rollup'
import v8 from 'v8'
import type {
  ManifestV3,
  WebAccessibleResourceById,
  WebAccessibleResourceByMatch,
} from './manifest'
import type { AcornLiteral, ManifestFiles } from './types'

export const _debug = (id: string) => debug('crx').extend(id)

export const structuredClone = <T>(obj: T): T => {
  return v8.deserialize(v8.serialize(obj))
}

export const createHash = (data: string, length = 5): string =>
  _hash('sha1')
    .update(data)
    .digest('base64')
    .replace(/[^A-Za-z0-9]/g, '')
    .slice(0, length)

export const isString = (x: unknown): x is string => typeof x === 'string'

type Falsy = false | 0 | '' | null | undefined
export const isTruthy = <T>(x: T | Falsy): x is T => !!x

export function isObject<T>(
  value: T,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
): value is Extract<T, Record<string, any>> {
  return Object.prototype.toString.call(value) === '[object Object]'
}

export const isResourceByMatch = (
  x: WebAccessibleResourceById | WebAccessibleResourceByMatch,
): x is WebAccessibleResourceByMatch => 'matches' in x

export async function manifestFiles(
  manifest: ManifestV3,
  options: fg.Options = {},
): Promise<ManifestFiles> {
  // JSON
  let locales: string[] = []
  if (manifest.default_locale)
    locales = await fg('_locales/**/messages.json', options)

  const rulesets =
    manifest.declarative_net_request?.rule_resources.flatMap(
      ({ path }) => path,
    ) ?? []

  const contentScripts = manifest.content_scripts?.flatMap(({ js }) => js) ?? []
  const contentStyles = manifest.content_scripts?.flatMap(({ css }) => css)
  const serviceWorker = manifest.background?.service_worker
  const htmlPages = htmlFiles(manifest)

  const icons = [
    Object.values(manifest.icons ?? {}) as string[],
    Object.values(manifest.action?.default_icon ?? {}) as string[],
  ].flat()

  return {
    contentScripts: [...new Set(contentScripts)].filter(isString),
    contentStyles: [...new Set(contentStyles)].filter(isString),
    html: htmlPages,
    icons: [...new Set(icons)].filter(isString),
    locales: [...new Set(locales)].filter(isString),
    rulesets: [...new Set(rulesets)].filter(isString),
    background: [serviceWorker].filter(isString),
  }
}

export function htmlFiles(manifest: ManifestV3): string[] {
  const files = [
    manifest.action?.default_popup,
    Object.values(manifest.chrome_settings_overrides ?? {}),
    manifest.devtools_page,
    manifest.options_page,
    manifest.options_ui?.page,
  ]
    .flat()
    .filter(isString)
    .sort()
  return [...new Set(files)]
}

export function decodeManifest(this: PluginContext, code: string): ManifestV3 {
  const tree = this.parse(code)
  let literal: AcornLiteral | undefined
  simple(tree, {
    Literal(n) {
      literal = n as AcornLiteral
    },
  })
  if (!literal) throw new Error('unable to parse manifest code')
  let result: ManifestV3 = JSON.parse(literal.value)
  if (typeof result === 'string') result = JSON.parse(result)
  return result
}

export function encodeManifest(manifest: ManifestV3): string {
  const json = JSON.stringify(JSON.stringify(manifest))
  return `export default ${json}`
}
