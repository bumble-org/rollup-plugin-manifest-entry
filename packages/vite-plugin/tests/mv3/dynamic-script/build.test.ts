import { build } from 'tests/runners'
import { testOutput } from 'tests/testOutput'
import { mockDate } from 'tests/helpers'

mockDate()

// TODO: handle main world scripts through web_accessible_resources

test('build fs output', async () => {
  const result = await build(__dirname)
  await testOutput(result)
})
