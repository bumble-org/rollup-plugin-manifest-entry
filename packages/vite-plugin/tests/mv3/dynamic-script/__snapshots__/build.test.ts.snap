// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`build fs output: _00 manifest.json 1`] = `
Object {
  "background": Object {
    "service_worker": "service-worker-loader.js",
    "type": "module",
  },
  "content_scripts": Array [
    Object {
      "js": Array [
        "assets/declared-script.ts-loader.hash.js",
      ],
      "matches": Array [
        "https://a.com/*",
        "http://b.com/*",
      ],
    },
  ],
  "manifest_version": 3,
  "name": "test extension",
  "version": "1.0.0",
  "web_accessible_resources": Array [
    Object {
      "matches": Array [
        "http://b.com/*",
        "https://a.com/*",
      ],
      "resources": Array [
        "assets/declared-script.ts.hash.js",
        "assets/main-world.ts.hash.js",
      ],
      "use_dynamic_url": true,
    },
    Object {
      "matches": Array [
        "http://*/*",
        "https://*/*",
      ],
      "resources": Array [
        "assets/dynamic-script.ts.hash.js",
      ],
      "use_dynamic_url": false,
    },
  ],
}
`;

exports[`build fs output: _01 output files 1`] = `
Array [
  "assets/background.ts.hash.js",
  "assets/declared-script.ts-loader.hash.js",
  "assets/declared-script.ts.hash.js",
  "assets/dynamic-script.ts-loader.hash.js",
  "assets/dynamic-script.ts.hash.js",
  "assets/main-world.ts.hash.js",
  "manifest.json",
  "service-worker-loader.js",
]
`;

exports[`build fs output: assets/background.ts.hash.js 1`] = `
"const import_meta = {};
var scriptFileName = \\"assets/dynamic-script.ts-loader.hash.js\\";
console.log(\\"background.ts\\");
chrome.action.onClicked.addListener((tab) => {
  if (tab.id)
    chrome.scripting.executeScript({
      files: [scriptFileName],
      target: { tabId: tab.id }
    });
});
"
`;

exports[`build fs output: assets/declared-script.ts.hash.js 1`] = `
"const import_meta = {};
var fileName = \\"assets/main-world.ts.hash.js\\";
console.log(\\"declared script\\");
const script = document.createElement(\\"script\\");
script.src = chrome.runtime.getURL(fileName);
script.type = \\"module\\";
document.head.prepend(script);
"
`;

exports[`build fs output: assets/declared-script.ts-loader.hash.js 1`] = `
"(function () {
  'use strict';

  (async () => {
    await import(
      /* @vite-ignore */
      chrome.runtime.getURL(\\"assets/declared-script.ts.hash.js\\")
    );
  })().catch(console.error);

})();
"
`;

exports[`build fs output: assets/dynamic-script.ts.hash.js 1`] = `
"console.log(\\"dynamic script\\");
"
`;

exports[`build fs output: assets/dynamic-script.ts-loader.hash.js 1`] = `
"(function () {
  'use strict';

  (async () => {
    await import(
      /* @vite-ignore */
      chrome.runtime.getURL(\\"assets/dynamic-script.ts.hash.js\\")
    );
  })().catch(console.error);

})();
"
`;

exports[`build fs output: assets/main-world.ts.hash.js 1`] = `
"console.log(\\"main world script\\");
"
`;

exports[`build fs output: service-worker-loader.js 1`] = `
"import './assets/background.ts.hash.js';
"
`;
